sycl/include/sycl/ext/intel/experimental/esimd/detail/test_proxy.hpp

```diff
@@ -0,0 +1,106 @@
+//
+// We are using non static data member initialization approach to force
+// the value required. Initialization will take place even if no
+// default/copy/move constructor of the test_proxy class was explcitly
+// called by any of the user-defined constructors of the proxy target
+#define __ESIMD_DECLARE_TEST_PROXY                                             \
+  esimd::detail::test_proxy M_testProxy = esimd::detail::test_proxy();
+
+// Declare the getter to access the proxy from the tests
+#define __ESIMD_DECLARE_TEST_PROXY_ACCESS                                      \
+  const auto &get_test_proxy() const { return M_testProxy; }
+
+// Test proxy will be handled in a proper way by default/implicit move
+// constructors and move operators.
+// Still the user-defined constructors or move operators should explicitly state
+// what to do with each of class atributes, so a proper wrapper required
+#define __esimd_move_test_proxy(other)                                         \
+  do {                                                                         \
+    M_testProxy = std::move(other.M_testProxy);                                \
+  } while (false);
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace ext {
+namespace intel {
+namespace experimental {
+namespace esimd {
+namespace detail {
+
+/// The test_proxy class.
+///
+/// This is a helper class for tests to differentiate between the copy
+/// constructor/assignment and the move constructor/assignment calls,
+/// as the copy constructor works as the default fallback for every case with
+/// move constructor disabled or not provided
+///
+/// The test proxy is enabled only if the __ESIMD_ENABLE_TEST_PROXY macro is
+/// defined.
+/// It is expected for the class with the test proxy (the class under test) to:
+/// - provide the get_test_proxy() method
+/// - properly handle moving the test_proxy member in user-defined move
+///   constructors and user-defined assignment operators
+///
+/// Therefore the following expression is expected to return `true` only if the
+/// move constructor or move operator was called for the instance of the class
+/// under test:
+///   instance.get_test_proxy().was_moved()
+///
+/// \ingroup sycl_esimd
+class test_proxy {
+  // Define the default value to use for every constructor
+  bool m_moved = false;
+
+public:
+  test_proxy() { __esimd_dbg_print(test_proxy()); }
+
+  test_proxy(const test_proxy &) {
+    __esimd_dbg_print(test_proxy(const test_proxy &other));
+  }
+  test_proxy(test_proxy &&) {
+    __esimd_dbg_print(test_proxy(test_proxy && other));
+    m_moved = true;
+  }
+  test_proxy &operator=(const test_proxy &) {
+    __esimd_dbg_print(test_proxy::operator=(const test_proxy &other));
+    return *this;
+  }
+  test_proxy &operator=(test_proxy &&) {
+    __esimd_dbg_print(test_proxy::operator=(test_proxy &&other));
+    m_moved = true;
+    return *this;
+  }
+  bool was_moved() const { return m_moved; }
+};
+
+} // namespace detail
```

Comment:
let's use separate namespace for the test stuff. `esimd::detail::test` ?

Reply:
Done by 0080b4e51b6467f73208ec54592a925c63147257, please take a look. I haven't added any subdir though, as I don't see any other cases when we would need the test stuff except this specific case.